doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1550 @@
+
+nat_entry_attr[4].id = SAI_NAT_ENTRY_ATTR_L4_DST_PORT;
+nat_entry_attr[4].value.u16 = 3000;
+
+nat_entry_attr[5].id = SAI_NAT_ENTRY_ATTR_ENABLE_PACKET_COUNT;
+nat_entry_attr[5].value.booldata = true;
+
+nat_entry_attr[6].id = SAI_NAT_ENTRY_ATTR_ENABLE_BYTE_COUNT;
+nat_entry_attr[6].value.booldata = true;
+
+attr_count = 7;
+
+memset(&dbl_nat_etnry, 0, sizeof(nat_etnry));
+
+dbl_nat_entry.data.key.src_ip = 200.200.200.1;
+dbl_nat_entry.data.mask.src_ip = 0xffffffff;
+dbl_nat_entry.data.key.proto = 17;  /* TCP */
+dbl_nat_entry.data.key.l4_src_port = 6000;
+dbl_nat_entry.data.mask.l4_src_port = 0xffff;
+dbl_nat_entry.data.key.dst_ip = 172.16.1.100;
+dbl_nat_entry.data.mask.dst_ip = 0xffffffff;
+dbl_nat_entry.data.key.l4_dst_port = 5000;
+dbl_nat_entry.data.mask.l4_dst_port = 0xffff;
+
+create_nat_entry(&dbl_nat_entry, attr_count, nat_entry_attr);
+
+/* Step 2: Create a Double NAPT entry for the reverse direction:
+* ------------------------------------------------------------ */
+sai_attribute_t nat_entry_attr[10];
+nat_entry_t dbl_nat_entry;
+
+nat_entry_attr[0].id = SAI_NAT_ENTRY_ATTR_NAT_TYPE;
+nat_entry_attr[0].value = SAI_NAT_TYPE_DOUBLE_NAT;
+
+nat_entry_attr[1].id = SAI_NAT_ENTRY_ATTR_SRC_IP;
+nat_entry_attr[1].value.u32 = 172.16.1.100;
+
+nat_entry_attr[2].id = SAI_NAT_ENTRY_ATTR_L4_SRC_PORT;
+nat_entry_attr[2].value.u16 = 5000;
+
+nat_entry_attr[3].id = SAI_NAT_ENTRY_ATTR_DST_IP;
+nat_entry_attr[3].value.u32 = 200.200.200.1;
+
+nat_entry_attr[4].id = SAI_NAT_ENTRY_ATTR_L4_DST_PORT;
+nat_entry_attr[4].value.u16 = 6000;
+
+nat_entry_attr[5].id = SAI_NAT_ENTRY_ATTR_ENABLE_PACKET_COUNT;
+nat_entry_attr[5].value.booldata = true;
+
+nat_entry_attr[6].id = SAI_NAT_ENTRY_ATTR_ENABLE_BYTE_COUNT;
+nat_entry_attr[6].value.booldata = true;
+
+attr_count = 7;
+
+memset(&dbl_nat_etnry, 0, sizeof(nat_etnry));
+
+dbl_nat_entry.data.key.src_ip = 200.200.200.100;
+dbl_nat_entry.data.mask.src_ip = 0xffffffff;
+dbl_nat_entry.data.key.proto = 17;  /* TCP */
+dbl_nat_entry.data.key.l4_src_port = 3000;
+dbl_nat_entry.data.mask.l4_src_port = 0xffff;
+dbl_nat_entry.data.key.dst_ip = 138.76.28.1;
+dbl_nat_entry.data.mask.dst_ip = 0xffffffff;
+dbl_nat_entry.data.key.l4_dst_port = 1024;
+dbl_nat_entry.data.mask.l4_dst_port = 0xffff;
+
+create_nat_entry(&dbl_nat_entry, attr_count, nat_entry_attr);
+
+```
+For example, to create a Counter entry per Zone per NAT type, NAT orchagent invokes the following SAI APIs with the necessary SAI attributes:
+
+```
+/* Step 1: Create a NAT Zone Counter Object for Source NAT Zone 1:
+* ------------------------------------------------------------------- */
+sai_attribute_t nat_zone_counter_attr[10];
+
+nat_zone_counter_attr[0].id = SAI_NAT_ZONE_COUNTER_ATTR_NAT_TYPE;
+nat_zone_counter _attr[0].value = SAI_NAT_TYPE_SOURCE_NAT;
+
+nat_zone_counter_attr[1].id = SAI_NAT_ZONE_COUNTER_ATTR_ZONE_ID;
+nat_zone_counter_attr[1].value.u32 = 1;
+
+nat_zone_counter_attr[2].id = SAI_NAT_ZONE_COUNTER_ATTR_ENABLE_TRANSLATION_NEEDED;
+nat_zone_co
```

Comment:
Could you please describe a bit more reconciliation in natsyncd?
Is the reconciliation done after a timer expires like in fpmsyncd warm restart case?
Based on what do you mark an entry as stale?
Do we need reconciliation here? What if we just restore and then dynamic NAT entries will be removed if no traffic hits them.

Reply:
Yes, reconciliation starts after the timer expires.
As explained in section 6 in this document:
The warmRestartAssist class is used to cache the data read from APP_DB.
Now the conntrack entries are read and pushed to warmRestartAssist class.
After the timer expires, the reconciliation is done by warmRestartAssist which removes those entries the stale entries from the APP_DB. Ideally all the entries should be in sync. In case for reason if a connrtack entry failed to be restored in the kernel, that entry is marked as stale and removed from app-db during reconciliation.